### ISSUE TITLE
rake package fails on exception reported from rubygems

### ISSUE THREAD

**Issue by mcorino** [seattlerb/hoe#85]
I recently upgraded to running `ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-linux-gnu]` with rubygems v. 2.7.3 and hoe v. 3.17.0 installed.
When executing `rake package` I now invariably get the following error reported for my gem projects:
```
rake aborted!
Errno::ENOENT: No such file or directory @ rb_file_s_stat - Gemfile
```
I never had this problem running my previous <ruby 2.5 installation.
Tracing rake reveals the error coming from the rubygems code:
```
rake aborted!
Errno::ENOENT: No such file or directory @ rb_file_s_stat - Gemfile
/usr/lib64/ruby/2.5.0/rubygems/package.rb:117:in `stat'
/usr/lib64/ruby/2.5.0/rubygems/package.rb:117:in `build'
/usr/lib64/ruby/2.5.0/rubygems/package_task.rb:118:in `block (3 levels) in define'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:112:in `when_writing'
/usr/lib64/ruby/2.5.0/rubygems/package_task.rb:117:in `block (2 levels) in define'
...
```
Any idea what's wrong?

**Comment by zenspider** [comment]
I would need the rest of that `...` part to really start to guess.

First steps:

1. Does this have any plugins involved (like `:git`?). Does deactivating them bypass this?
2. Do you have a file called Gemfile in your project?
3. Do you have Gemfile in your Manifest.txt file?
4. OR, are the plugins missing on the new installation? Plugins are soft errors if they aren't available. Run via `ruby -d -S rake` to see plugin failures (and LOTS of other noise, ignore 99% of it).

**Comment by zenspider** [comment]
I should also ask: what version of rubygems are you now on?

**Comment by mcorino** [comment]
I'm using rubygems v. 2.7.3.

Re 1: no, no plugins involved. I'm using a very basic Rakefile:
```
# -*- ruby -*-
require 'rubygems'
require 'hoe'

# Hoe.plugin :compiler
# Hoe.plugin :gem_prelude_sucks
# Hoe.plugin :inline
# Hoe.plugin :racc
# Hoe.plugin :rcov
# Hoe.plugin :rubyforge

Hoe.plugins.delete :test

Hoe.spec 'fuzz' do
  developer('Martin Corino', 'mcorino@remedy.nl')

  license "MIT"
  
  # self.rubyforge_name = 'fuzzx' # if different than 'fuzz'
end

# vim: syntax=ruby
```

Re 2: No
Re 3: No
Re 4: No plugins and no plugin failures

This is the full trace stack:
```
rake aborted!
Errno::ENOENT: No such file or directory @ rb_file_s_stat - Gemfile
/usr/lib64/ruby/2.5.0/rubygems/package.rb:117:in `stat'
/usr/lib64/ruby/2.5.0/rubygems/package.rb:117:in `build'
/usr/lib64/ruby/2.5.0/rubygems/package_task.rb:118:in `block (3 levels) in define'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:112:in `when_writing'
/usr/lib64/ruby/2.5.0/rubygems/package_task.rb:117:in `block (2 levels) in define'
/usr/lib64/ruby/2.5.0/fileutils.rb:122:in `chdir'
/usr/lib64/ruby/2.5.0/fileutils.rb:122:in `cd'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:38:in `chdir'
/usr/lib64/ruby/2.5.0/rubygems/package_task.rb:116:in `block in define'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:251:in `block in execute'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:251:in `each'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:251:in `execute'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:195:in `block in invoke_with_call_chain'
/usr/lib64/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:217:in `block in invoke_prerequisites'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:215:in `each'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:215:in `invoke_prerequisites'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:194:in `block in invoke_with_call_chain'
/usr/lib64/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:217:in `block in invoke_prerequisites'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:215:in `each'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:215:in `invoke_prerequisites'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:194:in `block in invoke_with_call_chain'
/usr/lib64/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:181:in `invoke'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:160:in `invoke_task'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:116:in `each'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:116:in `block in top_level'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:125:in `run_with_threads'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:110:in `top_level'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:83:in `block in run'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:186:in `standard_exception_handling'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:80:in `run'
/usr/lib64/ruby/gems/2.5.0/gems/rake-12.3.0/exe/rake:27:in `<top (required)>'
```

**Comment by mcorino** [comment]
Btw, I can reproduce this on a newly sow-ed project where I only uncomment the `developer` setting in the generated Rakefile.

**Comment by zenspider** [comment]
I can't. So you might have some customizations in your `~/.hoe_template/default` ?

```
% mv ~/.hoe_template{,.old}
% sow blah
...
% cd blah
% emacs Rakefile # fix developer line, add your delete :test line.
% rake
rake aborted!
Don't know how to build task 'default' (see --tasks)
```

So, maybe do the `mv` and `sow` above, then `diff -ru ~/.hoe_template{.old,}/default`

**Comment by mcorino** [comment]
The Errno::ENOENT is triggered by this method
```ruby
  def self.build spec, skip_validation=false
    gem_file = spec.file_name
    spec_file = spec.loaded_from
    spec_file = "Gemfile" if spec_file.nil? || spec_file.empty?

    package = new gem_file
    package.spec = spec
    package.build_time = File.stat(spec_file).mtime.to_i   # <== THIS LINE TRIGGERS ENOENT
    package.build skip_validation

    gem_file
  end
```
in `class Gem::Package` in file rubygems/package.rb.
`spec.loaded_from` is empty which seems logical to me as hoe does not seem to use a Gem spec from file (the `load_from` attribute of the spec being the path to the file the spec was loaded from).
My rubygems version however seems to expect the spec always to be loaded from a file and when the path is empty substitutes a default name and expects that file to exist unconditionally.
As an experiment I patched the method to assign `'Rakefile'` to `spec_file` and after that hoe worked beautifully again.
Now, if we only could get hoe to set the `loaded_from` attribute of the Gem spec to `Rakefile` (or whatever file `Hoe.spec {}` is declared in...

**Comment by mcorino** [comment]
I have now returned to using RVM to manage ruby versions and gemsets as I did previously and installed ruby 2.5.1 with rubygems 2.7.7.
In this environment the problem I had with hoe is gone.
On inspection of the Gem::Package#build method in rubygems it turned out the code causing the problems (retrieving mtime of the spec file) has been removed.
So I think it's safe to say the problem has been an intermittent regression in Rubygems which has been solved by now.
